File: whatsapp-sticker-uploader/app/src/main/java/com/stickeruploader/StickerPackLoader.java
package com.stickeruploader;

import android.os.Environment;
import android.util.Log;

import com.stickeruploader.models.Sticker;
import com.stickeruploader.models.StickerPack;

import java.io.File;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Loader che legge sticker dalla memoria esterna (WhatsApp Media folder)
 * Basato sul pattern ufficiale WhatsApp di sticker packs
 */
public class StickerPackLoader {

    private static final String TAG = "StickerPackLoader";
    private static final int STICKERS_PER_PACK = 30;
    private static final long MAX_STICKER_SIZE_BYTES = 100 * 1024L; // 100KB

    public static final File STICKER_DIR = new File(
            Environment.getExternalStorageDirectory(),
            "Android/media/com.whatsapp/WhatsApp/Media/WhatsApp Stickers"
    );

    public static List<StickerPack> loadAllPacks() {
        Log.d(TAG, "loadAllPacks() - caricando da: " + STICKER_DIR.getAbsolutePath());

        if (!STICKER_DIR.exists() || !STICKER_DIR.isDirectory()) {
            Log.w(TAG, "Directory non trovata: " + STICKER_DIR.getAbsolutePath());
            return new ArrayList<>();
        }

        // Leggi tutti i file .webp
        File[] files = STICKER_DIR.listFiles((dir, name) ->
                name.toLowerCase().endsWith(".webp")
        );

        if (files == null || files.length == 0) {
            Log.w(TAG, "Nessun file WebP trovato");
            return new ArrayList<>();
        }

        // Ordina per nome
        List<File> sortedFiles = new ArrayList<>(Arrays.asList(files));
        sortedFiles.sort((a, b) -> a.getName().compareTo(b.getName()));

        Log.d(TAG, "File trovati: " + sortedFiles.size());

        // Filtra file non validi
        List<File> validFiles = new ArrayList<>();
        for (File f : sortedFiles) {
            long size = f.length();
            // File deve essere > 0 e <= 100KB e non animato
            if (size > 0 && size <= MAX_STICKER_SIZE_BYTES && !isAnimatedWebP(f)) {
                validFiles.add(f);
                Log.d(TAG, "✅ File valido: " + f.getName() + " (" + size + " bytes)");
            } else {
                Log.d(TAG, "❌ File scartato: " + f.getName() + " (size=" + size + ", animated=" + isAnimatedWebP(f) + ")");
            }
        }

        Log.d(TAG, "File validi: " + validFiles.size());

        if (validFiles.isEmpty()) {
            return new ArrayList<>();
        }

        // Raggruppa in pack di 30 sticker
        List<StickerPack> packs = new ArrayList<>();
        for (int i = 0; i < validFiles.size(); i += STICKERS_PER_PACK) {
            int endIndex = Math.min(i + STICKERS_PER_PACK, validFiles.size());
            List<File> packFiles = validFiles.subList(i, endIndex);

            // WhatsApp richiede minimo 3 sticker per pack
            if (packFiles.size() < 3) {
                Log.d(TAG, "Pack scartato: meno di 3 sticker");
                continue;
            }

            int packNumber = packs.size() + 1;
            String packId = String.format("my_sticker_pack_%03d", packNumber);
            String packName = "New Stiker " + packNumber;

            // Crea lista sticker
            List<Sticker> stickers = new ArrayList<>();
            for (File f : packFiles) {
                stickers.add(new Sticker(f.getName(), Arrays.asList("😀"), ""));
            }

            // Primo file come tray image
            String trayImageFile = packFiles.get(0).getName();

            StickerPack pack = new StickerPack(
                    packId,
                    packName,
                    "Il mio dispositivo",
                    trayImageFile,
                    "",
                    "",
                    "",
                    "",
                    "1",
                    false,
                    false,
                    stickers
            );

            packs.add(pack);
            Log.d(TAG, "✅ Pack creato: " + packId + " con " + stickers.size() + " sticker");
        }

        Log.d(TAG, "Totale pack: " + packs.size());
        return packs;
    }

    /**
     * Rileva se un file WebP è animato controllando l'header
     */
    private static boolean isAnimatedWebP(File file) {
        if (file.length() < 20) return false;

        try (RandomAccessFile raf = new RandomAccessFile(file, "r")) {
            byte[] header = new byte[50];
            raf.read(header);

            // Cerca "ANIM" nei primi 50 byte
            for (int i = 0; i < header.length - 3; i++) {
                if (header[i] == 65 &&     // 'A'
                    header[i+1] == 78 &&   // 'N'
                    header[i+2] == 73 &&   // 'I'
                    header[i+3] == 77) {   // 'M'
                    return true;
                }
            }
            return false;
        } catch (Exception e) {
            Log.e(TAG, "Errore checking WebP animato: " + e.getMessage());
            return false;
        }
    }

    public static File getStickerFile(String fileName) {
        return new File(STICKER_DIR, fileName);
    }
}

File: whatsapp-sticker-uploader/app/src/main/java/com/stickeruploader/StickerPackLoader.kt
package com.stickeruploader

import android.graphics.BitmapFactory
import android.os.Environment
import com.stickeruploader.models.Sticker
import com.stickeruploader.models.StickerPack
import java.io.File
import java.io.RandomAccessFile

object StickerPackLoader {

    const val STICKERS_PER_PACK        = 30
    const val MIN_STICKERS_PER_PACK    = 3
    private const val MAX_ANIMATED_SIZE_BYTES = 500 * 1024L

    // Pack correnti - aggiornati ad ogni loadAllPacks(), usati dal ContentProvider
    var currentPacks: List<StickerPack> = emptyList()
        private set

    // Statistiche ultima validazione
    var lastValidCount: Int = 0
        private set
    var lastInvalidCount: Int = 0
        private set

    val STICKER_DIR: File by lazy {
        File(
            Environment.getExternalStorageDirectory(),
            "Android/media/com.whatsapp/WhatsApp/Media/WhatsApp Stickers"
        )
    }

    fun loadAllPacks(stickersPerPack: Int = STICKERS_PER_PACK): List<StickerPack> {
        if (!STICKER_DIR.exists() || !STICKER_DIR.isDirectory) return emptyList()

        val allFiles = STICKER_DIR.listFiles { file ->
            file.isFile && file.name.lowercase().endsWith(".webp")
        }?.sortedBy { it.name } ?: emptyList()

        var validCount = 0
        var invalidCount = 0

        val animatedFiles = mutableListOf<File>()

        for (file in allFiles) {
            val size = file.length()
            if (size <= 0L) continue

            val info = parseWebP(file)

            if (info.animated) {
                // Valida ogni sticker individualmente secondo i requisiti WhatsApp:
                // - dimensioni 512x512
                // - dimensione ≤ 500KB
                // - loop count = 0 (loop infinito)
                val sizeOk = size <= MAX_ANIMATED_SIZE_BYTES
                val dimsOk = info.width == 512 && info.height == 512
                val loopOk = info.loopCount == 0

                if (sizeOk && dimsOk && loopOk) {
                    animatedFiles.add(file)
                    validCount++
                } else {
                    invalidCount++
                    AppLogger.w("StickerPackLoader",
                        "Sticker scartato: ${file.name} " +
                        "[size=${size/1024}KB ok=$sizeOk, " +
                        "${info.width}x${info.height} ok=$dimsOk, " +
                        "loop=${info.loopCount} ok=$loopOk]")
                }
            }
        }

        lastValidCount = validCount
        lastInvalidCount = invalidCount

        val packs = mutableListOf<StickerPack>()
        val size = stickersPerPack.coerceIn(MIN_STICKERS_PER_PACK, STICKERS_PER_PACK)

        animatedFiles.chunked(size).forEachIndexed { index, files ->
            if (files.size < MIN_STICKERS_PER_PACK) return@forEachIndexed
            val num    = index + 1
            val packId = "my_anim_pack_%03d".format(num)
            packs.add(StickerPack(
                identifier    = packId,
                name          = "New Stiker Anim $num",
                publisher     = "Il mio dispositivo",
                trayImageFile = "${packId}_tray.webp",
                stickers      = files.map { Sticker(it.name, listOf("😀")) },
                isAnimated    = true
            ))
        }

        currentPacks = packs
        return packs
    }

    fun getStickerFile(fileName: String): File = File(STICKER_DIR, fileName)

    private fun is512x512(file: File): Boolean {
        return try {
            val opts = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            BitmapFactory.decodeFile(file.absolutePath, opts)
            opts.outWidth == 512 && opts.outHeight == 512
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Legge l'header WebP e restituisce: se è animato, larghezza e altezza del canvas, loop count.
     *
     * Struttura (offset dal byte 0):
     *   [0-3]   = "RIFF"
     *   [8-11]  = "WEBP"
     *   [12-15] = "VP8X"  (se presente)
     *   [20]    = flags (bit 0x02 = animazione)
     *   [24-26] = Canvas Width  - 1 (24-bit LE)
     *   [27-29] = Canvas Height - 1 (24-bit LE)
     *   [30-33] = "ANIM"  (se animato)
     *   [38-41] = background color
     *   [42-43] = loop count (0 = infinito, richiesto da WhatsApp)
     */
    private data class WebPInfo(val animated: Boolean, val width: Int, val height: Int, val loopCount: Int = -1)

    private fun parseWebP(file: File): WebPInfo {
        val unknown = WebPInfo(false, -1, -1)
        if (file.length() < 30) return unknown
        return try {
            RandomAccessFile(file, "r").use { raf ->
                val header = ByteArray(256)
                val read = raf.read(header)

                if (read < 12) return@use unknown

                // Firma RIFF...WEBP
                if (header[0] != 'R'.code.toByte() || header[1] != 'I'.code.toByte() ||
                    header[2] != 'F'.code.toByte() || header[3] != 'F'.code.toByte()) return@use unknown
                if (header[8] != 'W'.code.toByte() || header[9] != 'E'.code.toByte() ||
                    header[10] != 'B'.code.toByte() || header[11] != 'P'.code.toByte()) return@use unknown

                // Chunk VP8X presente: legge flag, dimensioni canvas e loop count
                if (read >= 30 &&
                    header[12] == 'V'.code.toByte() && header[13] == 'P'.code.toByte() &&
                    header[14] == '8'.code.toByte() && header[15] == 'X'.code.toByte()) {

                    val flags = header[20].toInt() and 0xFF
                    val animated = (flags and 0x02) != 0

                    val w = (header[24].toInt() and 0xFF) or
                            ((header[25].toInt() and 0xFF) shl 8) or
                            ((header[26].toInt() and 0xFF) shl 16)
                    val h = (header[27].toInt() and 0xFF) or
                            ((header[28].toInt() and 0xFF) shl 8) or
                            ((header[29].toInt() and 0xFF) shl 16)

                    // Legge loop count dal chunk ANIM (offset 30, se abbastanza dati)
                    var loopCount = -1
                    if (animated && read >= 44 &&
                        header[30] == 'A'.code.toByte() && header[31] == 'N'.code.toByte() &&
                        header[32] == 'I'.code.toByte() && header[33] == 'M'.code.toByte()) {
                        loopCount = (header[42].toInt() and 0xFF) or
                                    ((header[43].toInt() and 0xFF) shl 8)
                    }

                    return@use WebPInfo(animated, w + 1, h + 1, loopCount)
                }

                // Nessun VP8X: cerca ANIM come fallback
                var animResult: WebPInfo = unknown
                for (i in 0..read - 4) {
                    if (header[i]   == 'A'.code.toByte() &&
                        header[i+1] == 'N'.code.toByte() &&
                        header[i+2] == 'I'.code.toByte() &&
                        header[i+3] == 'M'.code.toByte()) {
                        val opts = BitmapFactory.Options().apply { inJustDecodeBounds = true }
                        BitmapFactory.decodeFile(file.absolutePath, opts)
                        val bw = if (opts.outWidth > 0) opts.outWidth else 512
                        val bh = if (opts.outHeight > 0) opts.outHeight else 512
                        animResult = WebPInfo(true, bw, bh, -1)
                        break
                    }
                }
                animResult
            }
        } catch (e: Exception) { unknown }
    }

    fun isAnimatedWebP(file: File): Boolean = parseWebP(file).animated
}
